test/extensions/filters/http/common/fuzz/uber_filter.cc

```diff
@@ -85,6 +91,16 @@ void UberFilterFuzzer::fuzz(
     HttpFilterFuzzer::accessLog(access_logger_.get(), stream_info_);
   }
 
+  if (with_main_event_loop_) {
+    // Ensure the event loop gets at least one event to end the test.
+    auto end_timer =
+        main_dispatcher_->createTimer([]() { ENVOY_LOG_MISC(trace, "server timer fired"); });
+    end_timer->enableTimer(std::chrono::milliseconds(5000));
```

Comment:
Is there any way to validate that the filter has completed? Rather than a hard five second loop it would make more sense to do a gentle spin-loop until either the filter completes or a max timeout is reached. As-is, this will pass on the error case "filter never completes", which is probably a worse case than an exception.

I'm thinking similar to this but inside a while (filter not completed) block, with say a 50ms timer, and an if (total time elapsed > threshold) FAIL() branch. That way the test can complete in less than 5 seconds (this method makes it take 5 seconds even if it could have completed immediately), and can detect additional types of failure.

But this does look like it would fix the specific failure case it was targeting.

Reply:
(Added bonus of this approach, you shouldn't need to specify whether the dispatcher is even required - if it's not then the filter should have completed the first time so it wouldn't even go into the loop.)

Reply:
The `main_dispatcher_->run(Event::DispatcherImpl::RunType::Block);` blocks until it has processed **one** event. That is, when the filter's event is already waiting, when we reach this line, it is processed immediately and the event-loop is terminated. The timer event is deleted when the `end_timer` object goes out of scope. Therefore, when the filter is quick, one does not note the 5 secs at all (they never occur). But when the filter takes forever, then its timeout is 5 seconds, because the event-loop will process the timer event.

One can of course FAIL in the timer callback, when that one is the reason for the event-loop to end. Is that desired?

Reply:
My concern here is that there's no guarantee that there's only supposed to be one dispatched event (I'd be surprised if that's the case for file_system_buffer) - if run only processes one event (that wasn't my understanding from the docs, I thought with block it processed until the queue is empty, but I could be wrong) then without a while loop you're likely just testing up to the file write, and not all the way through the filter. Ideally the test would be able to validate correctly with anywhere from zero to n dispatched events, without long-blocking, which is how it would be with my proposed loop (provided there is a way to validate that the filter terminated!) There might be a race if run consumes the queue until empty, in that if the new events are added quickly enough the filter will complete, but if the new event is not added until after run has terminated then the filter gets left with a queued event (after the other thread adds it) that is never processed.

Though yeah, if Block exits after either one event or at least one event, then an "until filter is terminated" loop with a 5 second fail timeout would be a solid solution.

To clarify: the *two* situations I'm concerned about are:
1. the dispatcher doesn't get an event because the filter doesn't do anything async. In the current implementation you either have to explicitly avoid the dispatcher block entirely, or suffer a 5 second delay. With a loop predicated on the filter terminating, you wouldn't enter the loop in this case because the filter completes before you reach the loop if there's no async action involved.
2. the dispatcher receives more than one event, asynchronously. In the current implementation the test passes, the filter is left in an incomplete state, and there is no validation that the filter completes. With a loop predicated on the filter terminating, all dispatched events get processed and if the filter still doesn't complete in 5 seconds the test can fail.

Reply:
I hope I understand your concerns. You are right `run(Block)` will process at least one event. More if possible, but I never have seen this to be deterministic or controlable. Therefore I prefer to regard `run(Block)` to process one event to keep it simple.

For the issue to figure when a filter has terminated: To my understanding are filters created in envoy and live until envoy terminates. They are kept in stand-by and are applied to streams whenever a stream is applicable. Therefore is there no termination criteria. To *my* understanding! Correct me, if I am wrong. The encoder- and decoder-filter both have some kind of `en-/decodeComplete()` method, which *may* be usable here. (I say "may" because I have been surprised by envoy many times, that methods do not as suspected.) To use these (virtual) methods one would have to overload the class for every filter to inject code into that process. With the number of filters available I don't think this is feasible. I have examined the `en-/decoder_callbacks`, which are mocks in the `filter_fuzz_test`. Those could be easily used to indicate a filter "termination", but I could not identify a callback, that would suit our needs (which does not mean none exists; the name could just be counter-intuitive, so that I never got the idea to use that specific callback for that task). Therefore I am kind of stuck here. 

Furthermore do it not know, what the requirements in the design of the `filter_fuzz_test` are/were. I.e. it is unclear, whether it is ok to instantiate a filter and terminated it in the middle of its processing or if the filter's whole task should be carried out in each run of the fuzzer. With the current implementation I doubt this is even possible, because the fuzzer knows nothing about the structure of messages a filter needs to be usable.

After all of this chatter, I am open for any ideas of how to make the processing more reliable. Not having found anything usable yet, I stick to the PR as is. Feel free to add your (code) proposals to make the fuzzer work better.

Reply:
Thx

Reply:
Actually, could you put in this loop, to help me to remember to get to it after this lands...

```
// TODO(ravenblackx): this can potentially exit the loop without the filter having completed.
// We should detect stream completion rather than just running one round of dispatcher.
```

Reply:
comment added.